### PR TITLE
Use company-mode for python auto-complete

### PR DESCRIPTION
**Pull request by bonega**
`ac-anaconda` seems broken.
It seems that most people in the Emacs community think that `company-mode`
is the way forward.

This commit removes `ac-anaconda` from the python layer.
I propose that we should use `company-mode` by default.

### CODE REVIEW

**Comment by bonega** [comment]
By broken I mean that it tries to complete by more or less everything in the local namespace.
Probably is fixable, but I rather see a good default.

File: contrib/lang/python/packages.el [hunk 1]
```diff
@@ -34,9 +27,8 @@ which require an initialization must be listed explicitly in the list.")
 
 (defun python/init-company-anaconda ()
   (use-package company-anaconda
-    :if (boundp 'company-backends)
-    :defer t
-    :init (add-to-list 'company-backends 'company-anaconda)))
+    :init (add-hook 'python-mode-hook 'company-mode)
```
Comment (syl20bnr) [inline contrib/lang/python/packages.el]: What is the motivation behind the trade of `:defer t` with a hook in `:init` ? `company-anaconda` has an autoload cookie.

**Comment by danielwuz** [comment]
Why do you like to make `company` as default for python contrib? It's my understanding that currently `spacemacs` still uses `auto-complete` as its completion system elsewhere. There is on-going discussion about using `company-mode` by default as replacement of `auto-complete`, but before that happens, I think it would be better to keep python contrib consistent with others.

Additionally, if you include `company-contrib` in your `dotspacemacs` file, `auto-complete` related packages will not be installed and `anaconda-company` works just fine. But with this PR, people who don't use `company-mode` won't be able to get auto complete feature when editing python files at all.

**Comment by syl20bnr** [comment]
Thank you very much @danielwuz, I did not catch this issue ;-)

I will soon generalize company contrib into a auto-completion layer by moving auto-complete there along with company. Then we will slowly move the packages to company front-end, I hope we will get all the company back ends we need to be able to deprecate auto-complete but if we can't it's not a big deal since we lazy load it anyway.

So basically turning on/off auto-completion will be just a matter of using a layer, I like this idea. Would be handy for some repro steps or bug hunting.

**Comment by bonega** [comment]
@danielwuz Because `auto-complete` doesn't work out of the box.
Working defaults for something this basic should always be provided.
Probably it can be fixed for `auto-complete`, but the whole Emacs community seems to be abandoning it.

With this PR you have to use `company` in python files.

Without most users will be stranded with a badly functioning `auto-complete`.
Expecting every new user to customize their `.spacemacs` is a bit of a stretch.
Also I expect that enabling `company-contrib` interacts badly with some other contrib?
I suppose that's why it is not default.

File: contrib/lang/python/packages.el (hunk 1 above)
Comment (bonega) [inline contrib/lang/python/packages.el]: Just my `elisp` inexperience showing off.

**Comment by danielwuz** [comment]
@bonega `company-contrib` isn't enabled by default is because of historical reason. That layer was added by people later after `spacemacs` already included `auto-complete`, but not because enabling it breaks some other contrib.

In the case you mentioned that "By broken I mean that it tries to complete by more or less everything in the local namespace.", I already explained it's because some backend of you `auto-complete` is sourcing those candidates in another PR. If you don't like it or consider it as a bug, you may modify `ac-sources` to meet your needs. In contrast, with this PR, people who don't use `company-mode` will get a completely broken user experience as completion system in spacemacs will not work for them at all.

I'm also a `company` user and would like to see it being the default completion system in spacemacs. but for each step we take towards that goal, we should take into consideration of various  user's preference as much as possible.

**Comment by bonega** [comment]
@danielwuz I don't understand really.
As far as I know `ac-anaconda` doesn't work for anyone out of the box.

Nobody seems interested in providing a fix for `ac-anaconda` and uses `company-contrib` instead.
Why just not deprecate `ac-anaconda` in favor of `company-anaconda`?

In this PR everyone automatically uses `company-anaconda` and therefore they will get sane completion.
Would you be satisfied if there was a variable for `python/use-ac-anaconda` or something?
With nil as default.

Are you telling me that the usage of `company` in this PR will conflict with the usage of `auto-complete` in other layers?

What I want: A default completion system that works without any config for new users.

**Comment by danielwuz** [comment]
@bonega Here's my point, with this update, will completion system work out of the box if a user does not include `company-contrib` in their `dotspacemacs` file? If it works, then I'm totally in favor of this move. 

Otherwise, if you believe there's an issue with `ac-anaconda`, the right way should be report to https://github.com/proofit404/ac-anaconda first, the author of `ac-anaconda` is the same one as `anaconda-mode`, and is very responsive.

**Comment by syl20bnr** [comment]
I see only people wanting the same thing here.

As @danielwuz explained this PR won't work out of the box because it makes the assumption that the users use the company-mode layer.

As I said not very clearly in the previous comment there is one missing step before being able to switch the default completion back-end. 

This change is not so obvious as it seems, it is not only about the auto-completion, it also about on the fly checking etc... (any layer that introduces some sort of layer dependencies).

I will work on it soon. It the meantime I close this PR.
@bonega I'll contact you when it is time to switch the Python completion back-end so you will be able to resubmit your contribution.

**Comment by bonega** [comment]
Huh, I am looking at a running version right now that doesn't include `company-contrib`.
I suppose that I have been fooled by some cached code?

**Comment by syl20bnr** [comment]
oh you are right, company is installed as a dependency.
well I still prefer to wait for the huge cleaning required in auto-completion stuff in spacemacs. I just had a conversation on the chat about AC and company and I can see a great mess in the actual configurations. I make this a priority task so I should be able to work on it shortly and have a clean basis to enable the right backend (company) for python and the other modes.

**Comment by bonega** [comment]
All right.
It will be a good day when autocomplete can be deprecated.
